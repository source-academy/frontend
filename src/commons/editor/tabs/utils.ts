import path from 'path';

/**
 * Returns the shortest file paths that is uniquely identifiable among
 * all open editor tabs. This is similar to how most code editors available
 * handle the displaying of file names.
 *
 * For example, if there are 3 open editor tabs where the file names are
 * exactly the same, we would need to display more of the file path to be
 * able to distinguish which editor tab corresponds to which file. Given
 * the following absolute file paths:
 * - /a.js
 * - /dir1/a.js
 * - /dir1/dir2/a.js
 * The shortest unique file paths will be:
 * - /a.js
 * - /dir1/a.js
 * - /dir2/a.js
 *
 * @param originalFilePaths The file paths to be shortened.
 */
export const getShortestUniqueFilePaths = (originalFilePaths: string[]): string[] => {
  // Store the unique shortened file paths as a mapping from the original file paths
  // to the shortened file paths. This is necessary because the output of this function
  // must preserve the original ordering of file paths.
  const originalToUniqueShortenedFilePaths: Record<string, string> = {};
  // Split each original file path into path segments and store the mapping from file
  // path to path segments for O(1) lookup. Since we only deal with the BrowserFS file
  // system, the path separator will always be '/'.
  const filePathSegments: Record<string, string[]> = originalFilePaths.reduce(
    (segments, filePath) => ({
      ...segments,
      // It is necessary to remove empty segments to deal with the very first '/' in
      // file paths.
      [filePath]: filePath.split(path.sep).filter(segment => segment !== '')
    }),
    {}
  );

  for (
    let numOfPathSegments = 1;
    // Keep looping while some original file paths have yet to be shortened.
    Object.keys(originalToUniqueShortenedFilePaths).length < originalFilePaths.length;
    numOfPathSegments++
  ) {
    // Based on the number of path segments for the iteration, we construct the
    // shortened file path. We then store the mapping from the shortened file path
    // to any original file path which transforms into it.
    const shortenedToOriginalFilePaths: Record<string, string[]> = Object.entries(
      filePathSegments
    ).reduce((filePaths, [originalFilePath, filePathSegments]) => {
      // Note that if there are fewer path segments than the number being sliced,
      // all of the path segments will be returned without error.
      const shortenedFilePath = '/' + filePathSegments.slice(-numOfPathSegments).join('/');
      return {
        ...filePaths,
        [shortenedFilePath]: (filePaths[shortenedFilePath] ?? []).concat(originalFilePath)
      };
    }, {});
    // Each shortened file path that only has a single corresponding original file
    // path is added to the unique shortened file paths record and their entry in
    // the file path segments record is removed to prevent further processing.
    Object.entries(shortenedToOriginalFilePaths).forEach(
      ([shortenedFilePath, originalFilePaths]) => {
        if (originalFilePaths.length > 1) {
          return;
        }

        const originalFilePath = originalFilePaths[0];
        originalToUniqueShortenedFilePaths[originalFilePath] = shortenedFilePath;
        // Remove the file path's segments from the next iteration.
        delete filePathSegments[originalFilePath];
      }
    );
  }

  // Finally, we retrieve the unique shortened file paths while preserving the ordering
  // of file paths.
  return originalFilePaths.map(filePath => originalToUniqueShortenedFilePaths[filePath]);
};
