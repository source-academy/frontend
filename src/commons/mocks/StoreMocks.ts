import _ from 'lodash';
import { DeepPartial, Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultAchievement,
  defaultApplication,
  defaultDashboard,
  defaultFileSystem,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore(overrides?: DeepPartial<OverallState>): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    achievement: defaultAchievement,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession,
    fileSystem: defaultFileSystem
  };
  return createStore(_.merge(state, overrides));
}
