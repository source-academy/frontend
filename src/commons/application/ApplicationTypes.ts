import { Chapter, Language, SourceError, Variant } from 'js-slang/dist/types';

import { AcademyState } from '../../features/academy/AcademyTypes';
import { AchievementState } from '../../features/achievement/AchievementTypes';
import { DashboardState } from '../../features/dashboard/DashboardTypes';
import { Grading } from '../../features/grading/GradingTypes';
import { PlaygroundState } from '../../features/playground/PlaygroundTypes';
import { PlaybackStatus, RecordingStatus } from '../../features/sourceRecorder/SourceRecorderTypes';
import { WORKSPACE_BASE_PATHS } from '../../pages/fileSystem/createInBrowserFileSystem';
import { Assessment } from '../assessment/AssessmentTypes';
import { FileSystemState } from '../fileSystem/FileSystemTypes';
import Constants from '../utils/Constants';
import { createContext } from '../utils/JsSlangHelper';
import {
  DebuggerContext,
  WorkspaceLocation,
  WorkspaceManagerState,
  WorkspaceState
} from '../workspace/WorkspaceTypes';
import { ExternalLibraryName } from './types/ExternalTypes';
import { SessionState } from './types/SessionTypes';

export type OverallState = {
  readonly academy: AcademyState;
  readonly achievement: AchievementState;
  readonly application: ApplicationState;
  readonly playground: PlaygroundState;
  readonly session: SessionState;
  readonly workspaces: WorkspaceManagerState;
  readonly dashboard: DashboardState;
  readonly fileSystem: FileSystemState;
};

export type ApplicationState = {
  readonly environment: ApplicationEnvironment;
};

export type Story = {
  story: string;
  playStory: boolean;
};

export type GameState = {
  collectibles: { [id: string]: string };
  completed_quests: string[];
};

/**
 * An output while the program is still being run in the interpreter. As a
 * result, there are no return values or SourceErrors yet. However, there could
 * have been calls to display (console.log) that need to be printed out.
 */
export type RunningOutput = {
  type: 'running';
  consoleLogs: string[];
};

/**
 * An output which reflects the program which the user had entered. Not a true
 * Output from the interpreter, but simply there to let he user know what had
 * been entered.
 */
export type CodeOutput = {
  type: 'code';
  value: string;
};

/**
 * An output which represents a program being run successfully, i.e. with a
 * return value at the end. A program can have either a return value, or errors,
 * but not both.
 */
export type ResultOutput = {
  type: 'result';
  value: any;
  consoleLogs: string[];
  runtime?: number;
  isProgram?: boolean;
};

/**
 * An output which represents a program being run unsuccessfully, i.e. with
 * errors at the end. A program can have either a return value, or errors, but
 * not both.
 */
export type ErrorOutput = {
  type: 'errors';
  errors: SourceError[];
  consoleLogs: string[];
};

export type InterpreterOutput = RunningOutput | CodeOutput | ResultOutput | ErrorOutput;

export enum ApplicationEnvironment {
  Development = 'development',
  Production = 'production',
  Test = 'test'
}

export enum Role {
  Student = 'student',
  Staff = 'staff',
  Admin = 'admin'
}

/**
 * Defines the languages available for use on Source Academy,
 * including Source sublanguages and other languages e.g. full JS.
 * For external libraries, see ExternalLibrariesTypes.ts
 */
export interface SALanguage extends Language {
  displayName: string;
}

const variantDisplay: Map<Variant, string> = new Map([
  [Variant.TYPED, 'Typed'],
  [Variant.WASM, 'WebAssembly'],
  [Variant.NON_DET, 'Non-Det'],
  [Variant.CONCURRENT, 'Concurrent'],
  [Variant.LAZY, 'Lazy'],
  [Variant.GPU, 'GPU'],
  [Variant.NATIVE, 'Native'],
  [Variant.EXPLICIT_CONTROL, 'Explicit-Control']
]);

export const fullJSLanguage: SALanguage = {
  chapter: Chapter.FULL_JS,
  variant: Variant.DEFAULT,
  displayName: 'full JavaScript'
};

export const fullTSLanguage: SALanguage = {
  chapter: Chapter.FULL_TS,
  variant: Variant.DEFAULT,
  displayName: 'full TypeScript'
};

export const htmlLanguage: SALanguage = {
  chapter: Chapter.HTML,
  variant: Variant.DEFAULT,
  displayName: 'HTML'
};

export const styliseSublanguage = (chapter: Chapter, variant: Variant = Variant.DEFAULT) => {
  switch (chapter) {
    case Chapter.FULL_JS:
      return fullJSLanguage.displayName;
    case Chapter.FULL_TS:
      return fullTSLanguage.displayName;
    case Chapter.HTML:
      return htmlLanguage.displayName;
    default:
      return `Source \xa7${chapter}${
        variantDisplay.has(variant) ? ` ${variantDisplay.get(variant)}` : ''
      }`;
  }
};

export const sublanguages: Language[] = [
  { chapter: Chapter.SOURCE_1, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_1, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_1, variant: Variant.WASM },
  { chapter: Chapter.SOURCE_1, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_1, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_2, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_2, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_2, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_2, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_3, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_3, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_3, variant: Variant.CONCURRENT },
  { chapter: Chapter.SOURCE_3, variant: Variant.NON_DET },
  { chapter: Chapter.SOURCE_3, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_4, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_4, variant: Variant.GPU },
  { chapter: Chapter.SOURCE_4, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.EXPLICIT_CONTROL }
];

export const sourceLanguages: SALanguage[] = sublanguages.map(sublang => {
  return {
    ...sublang,
    displayName: styliseSublanguage(sublang.chapter, sublang.variant)
  };
});

export const defaultLanguages = sourceLanguages.filter(
  sublang => sublang.variant === Variant.DEFAULT
);

export const variantLanguages = sourceLanguages.filter(
  sublang => sublang.variant !== Variant.DEFAULT
);

export const isSourceLanguage = (chapter: Chapter) =>
  [Chapter.SOURCE_1, Chapter.SOURCE_2, Chapter.SOURCE_3, Chapter.SOURCE_4].includes(chapter);

const currentEnvironment = (): ApplicationEnvironment => {
  switch (process.env.NODE_ENV) {
    case 'development':
      return ApplicationEnvironment.Development;
    case 'production':
      return ApplicationEnvironment.Production;
    default:
      return ApplicationEnvironment.Test;
  }
};

export const defaultAcademy: AcademyState = {
  gameCanvas: undefined
};

export const defaultApplication: ApplicationState = {
  environment: currentEnvironment()
};

export const defaultDashboard: DashboardState = {
  gradingSummary: {
    cols: [],
    rows: []
  }
};

export const defaultAchievement: AchievementState = {
  achievements: [],
  goals: [],
  users: [],
  assessmentOverviews: []
};

export const defaultPlayground: PlaygroundState = {
  githubSaveInfo: { repoName: '', filePath: '' }
};

export const defaultEditorValue = '// Type your program in here!';

/**
 * Create a default IWorkspaceState for 'resetting' a workspace.
 * Takes in parameters to set the js-slang library and chapter.
 *
 * @param workspaceLocation the location of the workspace, used for context
 */
export const createDefaultWorkspace = (workspaceLocation: WorkspaceLocation): WorkspaceState => ({
  autogradingResults: [],
  context: createContext<WorkspaceLocation>(
    Constants.defaultSourceChapter,
    [],
    workspaceLocation,
    Constants.defaultSourceVariant
  ),
  isFolderModeEnabled: false,
  activeEditorTabIndex: 0,
  editorTabs: [
    {
      filePath: ['playground', 'sicp'].includes(workspaceLocation)
        ? getDefaultFilePath(workspaceLocation)
        : undefined,
      value: ['playground', 'sourcecast', 'githubAssessments'].includes(workspaceLocation)
        ? defaultEditorValue
        : '',
      highlightedLines: [],
      breakpoints: []
    }
  ],
  programPrependValue: '',
  programPostpendValue: '',
  editorSessionId: '',
  isEditorReadonly: false,
  editorTestcases: [],
  externalLibrary: ExternalLibraryName.NONE,
  execTime: 1000,
  output: [],
  replHistory: {
    browseIndex: null,
    records: [],
    originalValue: ''
  },
  replValue: '',
  sharedbConnected: false,
  stepLimit: 1000,
  globals: [],
  isEditorAutorun: false,
  isRunning: false,
  isDebugging: false,
  enableDebugging: true,
  debuggerContext: {} as DebuggerContext
});

const defaultFileName = 'program.js';
export const getDefaultFilePath = (workspaceLocation: WorkspaceLocation) =>
  `${WORKSPACE_BASE_PATHS[workspaceLocation]}/${defaultFileName}`;

export const defaultWorkspaceManager: WorkspaceManagerState = {
  assessment: {
    ...createDefaultWorkspace('assessment'),
    currentAssessment: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  grading: {
    ...createDefaultWorkspace('grading'),
    currentSubmission: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  playground: {
    ...createDefaultWorkspace('playground'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('playground'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  sourcecast: {
    ...createDefaultWorkspace('sourcecast'),
    audioUrl: '',
    codeDeltasToApply: null,
    currentPlayerTime: 0,
    description: null,
    inputToApply: null,
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    playbackDuration: 0,
    playbackStatus: PlaybackStatus.paused,
    sourcecastIndex: null,
    title: null,
    uid: null
  },
  sourcereel: {
    ...createDefaultWorkspace('sourcereel'),
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    recordingStatus: RecordingStatus.notStarted,
    timeElapsedBeforePause: 0,
    timeResumed: 0
  },
  sicp: {
    ...createDefaultWorkspace('sicp'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('sicp'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  githubAssessment: {
    ...createDefaultWorkspace('githubAssessment'),
    hasUnsavedChanges: false
  }
};

export const defaultSession: SessionState = {
  courses: [],
  group: null,
  gameState: {
    completed_quests: [],
    collectibles: {}
  },
  xp: 0,
  allUserXp: undefined,
  story: {
    story: '',
    playStory: false
  },
  assessments: new Map<number, Assessment>(),
  assessmentOverviews: undefined,
  agreedToResearch: undefined,
  sessionId: Date.now(),
  githubOctokitObject: { octokit: undefined },
  gradingOverviews: undefined,
  gradings: new Map<number, Grading>(),
  notifications: []
};

export const defaultFileSystem: FileSystemState = {
  inBrowserFileSystem: null
};

export const defaultState: OverallState = {
  academy: defaultAcademy,
  achievement: defaultAchievement,
  application: defaultApplication,
  dashboard: defaultDashboard,
  playground: defaultPlayground,
  session: defaultSession,
  workspaces: defaultWorkspaceManager,
  fileSystem: defaultFileSystem
};
