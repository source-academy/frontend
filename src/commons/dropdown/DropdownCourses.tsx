import { Classes, Dialog, HTMLSelect } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';

import { Role } from '../application/ApplicationTypes';
import { UserCourse } from '../application/types/SessionTypes';
import { history } from '../utils/HistoryHelper';

type DialogProps = {
  isOpen: boolean;
  onClose: () => void;
  courses: UserCourse[];
  courseId?: number;
};

const DropdownCourses: React.FC<DialogProps> = props => {
  const { courses } = props;

  const options = courses.map(course => ({
    value: course.courseId,
    label: course.courseName.concat(!course.viewable ? ' - disabled' : ''),
    disabled: !course.viewable && course.role !== Role.Admin
  }));

  const onChangeHandler = (e: React.ChangeEvent<HTMLSelectElement>) => {
    history.push(`/courses/${e.currentTarget.value}`);
    props.onClose();
  };

  return (
    <Dialog
      className="my-courses"
      icon={IconNames.PROPERTIES}
      isCloseButtonShown={true}
      isOpen={props.isOpen}
      onClose={props.onClose}
      title="My Courses"
    >
      <div className={Classes.DIALOG_BODY}>
        <div>Select a course to switch to:</div>
        <HTMLSelect
          value={props.courseId}
          options={options}
          fill
          onChange={onChangeHandler}
          disabled={courses.length <= 1}
        />
      </div>
    </Dialog>
  );
};

export default DropdownCourses;
