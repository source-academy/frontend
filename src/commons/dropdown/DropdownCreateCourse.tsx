import {
  Button,
  Classes,
  Dialog,
  FormGroup,
  HTMLSelect,
  InputGroup,
  Switch,
  Tab,
  Tabs,
  Text,
  TextArea
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Chapter, Variant } from 'js-slang/dist/types';
import * as React from 'react';

import { CourseHelpTextEditorTab } from '../../pages/academy/adminPanel/subcomponents/CourseConfigPanel';
import { sourceLanguages } from '../application/ApplicationTypes';
import { UpdateCourseConfiguration } from '../application/types/SessionTypes';
import Markdown from '../Markdown';
import { showWarningMessage } from '../utils/NotificationsHelper';

type DialogProps = {
  isOpen: boolean;
  onClose: () => void;
  handleCreateCourse: (courseConfig: UpdateCourseConfiguration) => void;
};

const DropdownCreateCourse: React.FC<DialogProps> = props => {
  const [courseConfig, setCourseConfig] = React.useState<UpdateCourseConfiguration>({
    courseName: '',
    courseShortName: '',
    viewable: true,
    enableGame: true,
    enableAchievements: true,
    enableSourcecast: true,
    sourceChapter: Chapter.SOURCE_1,
    sourceVariant: Variant.DEFAULT,
    moduleHelpText: ''
  });

  const [courseHelpTextSelectedTab, setCourseHelpTextSelectedTab] =
    React.useState<CourseHelpTextEditorTab>(CourseHelpTextEditorTab.WRITE);

  const sourceChapterOptions = [
    { value: Chapter.SOURCE_1 },
    { value: Chapter.SOURCE_2 },
    { value: Chapter.SOURCE_3 },
    { value: Chapter.SOURCE_4 }
  ];

  const sourceVariantOptions = (chapter: Chapter) =>
    sourceLanguages
      .filter(e => e.chapter === chapter)
      .map(e => {
        return {
          label: e.variant.replace(/^\w/, c => c.toUpperCase()),
          value: e.variant
        };
      });

  const submitHandler = () => {
    // Validate that courseName is not an empty string
    if (courseConfig.courseName === '') {
      showWarningMessage('Course Name cannot be empty!');
      return;
    }
    props.handleCreateCourse(courseConfig);
    props.onClose();
  };

  const onChangeTabs = React.useCallback(
    (
      newTabId: CourseHelpTextEditorTab,
      prevTabId: CourseHelpTextEditorTab,
      event: React.MouseEvent<HTMLElement>
    ) => {
      if (newTabId === prevTabId) {
        return;
      }
      setCourseHelpTextSelectedTab(newTabId);
    },
    [setCourseHelpTextSelectedTab]
  );

  return (
    <Dialog
      className="create-course"
      icon={IconNames.ADD}
      isCloseButtonShown={true}
      canOutsideClickClose={false}
      isOpen={props.isOpen}
      onClose={props.onClose}
      title="Create Course"
    >
      <div className={Classes.DIALOG_BODY}>
        <div>
          <b>Create your own Source Academy course and manage your own learners!</b>
        </div>
        <br />
        <FormGroup
          helperText="Please enter the course name that will be used for course selection"
          label="Course Name"
          labelInfo="(* required)"
          labelFor="courseName"
        >
          <InputGroup
            id="courseName"
            value={courseConfig.courseName}
            onChange={e =>
              setCourseConfig({
                ...courseConfig,
                courseName: e.target.value
              })
            }
          />
        </FormGroup>
        <FormGroup
          helperText="Usually the module code of the course. This will be displayed on the top left."
          label={'Course Short Name'}
          labelInfo="(optional)"
          labelFor="courseShortName"
        >
          <InputGroup
            id="courseShortName"
            value={courseConfig.courseShortName}
            onChange={e =>
              setCourseConfig({
                ...courseConfig,
                courseShortName: e.target.value
              })
            }
          />
        </FormGroup>
        <FormGroup
          helperText="The module help text will be used in the course help dialog."
          labelFor="moduleHelpText"
        >
          <Text tagName="span">Module Help Text&nbsp;</Text>
          <Text tagName="span" className="optional-text">
            (optional)
          </Text>
          <Tabs
            selectedTabId={courseHelpTextSelectedTab}
            onChange={onChangeTabs}
            className="module-help-text-tabs"
          >
            <Tab id={CourseHelpTextEditorTab.WRITE} title="Write" />
            <Tab id={CourseHelpTextEditorTab.PREVIEW} title="Preview" />
          </Tabs>
          {courseHelpTextSelectedTab === CourseHelpTextEditorTab.WRITE && (
            <TextArea
              id="moduleHelpText"
              className="input-textarea"
              fill={true}
              value={courseConfig.moduleHelpText}
              onChange={e =>
                setCourseConfig({
                  ...courseConfig,
                  moduleHelpText: e.target.value
                })
              }
            />
          )}
          {courseHelpTextSelectedTab === CourseHelpTextEditorTab.PREVIEW && (
            <div className="input-markdown">
              <Markdown content={courseConfig.moduleHelpText || ''} openLinksInNewWindow />
            </div>
          )}
        </FormGroup>

        <div className="boolean-container">
          <div>
            <Switch
              checked={courseConfig.viewable}
              inline
              label="Viewable"
              onChange={e =>
                setCourseConfig({
                  ...courseConfig,
                  viewable: (e.target as HTMLInputElement).checked
                })
              }
            />
            <Switch
              checked={courseConfig.enableAchievements}
              inline
              label="Enable Achievements"
              onChange={e =>
                setCourseConfig({
                  ...courseConfig,
                  enableAchievements: (e.target as HTMLInputElement).checked
                })
              }
            />
          </div>
          <div>
            <Switch
              checked={courseConfig.enableGame}
              inline
              label="Enable Game"
              onChange={e =>
                setCourseConfig({
                  ...courseConfig,
                  enableGame: (e.target as HTMLInputElement).checked
                })
              }
            />

            <Switch
              checked={courseConfig.enableSourcecast}
              inline
              label="Enable Sourcecast"
              onChange={e =>
                setCourseConfig({
                  ...courseConfig,
                  enableSourcecast: (e.target as HTMLInputElement).checked
                })
              }
            />
          </div>
        </div>
        <div>
          <FormGroup
            label="Default Source Chapter"
            labelInfo="(configurable later on)"
            labelFor="source-chapter"
          >
            <HTMLSelect
              id="source-chapter"
              options={sourceChapterOptions}
              value={courseConfig.sourceChapter}
              onChange={e => {
                setCourseConfig({
                  ...courseConfig,
                  sourceChapter: parseInt(e.target.value)
                });
              }}
              fill
            />
          </FormGroup>
          <FormGroup
            label="Default Source Variant"
            labelInfo="(configurable later on)"
            labelFor="source-variant"
          >
            <HTMLSelect
              id="source-variant"
              options={sourceVariantOptions(courseConfig.sourceChapter!)}
              value={courseConfig.sourceVariant}
              onChange={e => {
                setCourseConfig({
                  ...courseConfig,
                  sourceVariant: e.target.value as Variant
                });
              }}
              fill
            />
          </FormGroup>
        </div>
        <div className="create-course-button-container">
          <Button text="Create Course" onClick={submitHandler} />
        </div>
      </div>
    </Dialog>
  );
};

export default DropdownCreateCourse;
