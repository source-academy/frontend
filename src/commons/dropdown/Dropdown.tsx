import { Menu, MenuItem, Position } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Popover2 } from '@blueprintjs/popover2';
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';

import { logOut } from '../application/actions/CommonsActions';
import ControlButton from '../ControlButton';
import Profile from '../profile/Profile';
import { useSession } from '../utils/Hooks';
import DropdownAbout from './DropdownAbout';
import DropdownCourses from './DropdownCourses';
import DropdownCreateCourse from './DropdownCreateCourse';
import DropdownHelp from './DropdownHelp';
import DropdownSettings from './DropdownSettings';

const Dropdown: React.FC = () => {
  const [isSettingsOpen, setIsSettingsOpen] = useState(false);
  const [isAboutOpen, setIsAboutOpen] = useState(false);
  const [isHelpOpen, setIsHelpOpen] = useState(false);
  const [isProfileOpen, setIsProfileOpen] = useState(false);
  const [isMyCoursesOpen, setIsMyCoursesOpen] = useState(false);
  const [isCreateCourseOpen, setIsCreateCourseOpen] = useState(false);

  const { isLoggedIn, name, courses, courseId } = useSession();
  const dispatch = useDispatch();
  const handleLogOut = () => dispatch(logOut());

  const toggleSettingsOpen = () => {
    setIsSettingsOpen(oldValue => !oldValue);
  };
  const toggleAboutOpen = () => {
    setIsAboutOpen(oldValue => !oldValue);
  };
  const toggleHelpOpen = () => setIsHelpOpen(oldValue => !oldValue);
  const toggleProfileOpen = () => setIsProfileOpen(oldValue => !oldValue);
  const toggleMyCoursesOpen = () => setIsMyCoursesOpen(oldValue => !oldValue);
  const toggleCreateCourseOpen = () => setIsCreateCourseOpen(oldValue => !oldValue);

  const profile =
    isLoggedIn && courseId != null ? (
      // Name is defined when user is logged in
      <MenuItem icon={IconNames.USER} onClick={toggleProfileOpen} text={titleCase(name!)} />
    ) : null;

  const myCourses = isLoggedIn ? (
    <MenuItem icon={IconNames.PROPERTIES} onClick={toggleMyCoursesOpen} text="My Courses" />
  ) : null;

  const createCourse = isLoggedIn ? (
    <MenuItem icon={IconNames.ADD} onClick={toggleCreateCourseOpen} text="Create Course" />
  ) : null;

  const logout = isLoggedIn ? (
    <MenuItem icon={IconNames.LOG_OUT} text="Logout" onClick={handleLogOut} />
  ) : null;

  const menu = (
    <Menu>
      {profile}
      {myCourses}
      {createCourse}
      <MenuItem icon={IconNames.COG} onClick={toggleSettingsOpen} text="Settings" />
      <MenuItem icon={IconNames.HELP} onClick={toggleAboutOpen} text="About" />
      <MenuItem icon={IconNames.ERROR} onClick={toggleHelpOpen} text="Help" />
      {logout}
    </Menu>
  );

  return (
    <>
      <Popover2 content={menu} inheritDarkTheme={false} placement={Position.BOTTOM}>
        <ControlButton icon={IconNames.CARET_DOWN} />
      </Popover2>
      <DropdownSettings isOpen={isSettingsOpen} onClose={toggleSettingsOpen} />
      <DropdownAbout isOpen={isAboutOpen} onClose={toggleAboutOpen} />
      <DropdownHelp isOpen={isHelpOpen} onClose={toggleHelpOpen} />
      {isLoggedIn ? (
        <>
          <DropdownCourses
            isOpen={isMyCoursesOpen}
            onClose={toggleMyCoursesOpen}
            courses={courses}
            courseId={courseId}
          />
          <DropdownCreateCourse isOpen={isCreateCourseOpen} onClose={toggleCreateCourseOpen} />
          <Profile isOpen={isProfileOpen} onClose={toggleProfileOpen} />
        </>
      ) : null}
    </>
  );
};

const titleCase = (str: string) =>
  str.replace(/\w\S*/g, wrd => wrd.charAt(0).toUpperCase() + wrd.substr(1).toLowerCase());

export default Dropdown;
