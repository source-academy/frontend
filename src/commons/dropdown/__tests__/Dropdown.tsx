import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { Store } from 'redux';
import { OverallState } from 'src/commons/application/ApplicationTypes';
import { UserCourse } from 'src/commons/application/types/SessionTypes';
import { mockInitialStore } from 'src/commons/mocks/StoreMocks';
import { renderTree, renderTreeJson } from 'src/commons/utils/TestUtils';
import { EditorBinding, WorkspaceSettingsContext } from 'src/commons/WorkspaceSettingsContext';

import Profile from '../../profile/Profile';
import Dropdown from '../Dropdown';
import DropdownCourses from '../DropdownCourses';
import DropdownCreateCourse from '../DropdownCreateCourse';

const getMockedStore = ({
  name,
  courses,
  courseId
}: {
  name?: string;
  courses: UserCourse[];
  courseId: number;
}) =>
  mockInitialStore({
    session: {
      name,
      courses,
      courseId
    }
  });

const getElement = (mockStore: Store<OverallState>) => (
  <Provider store={mockStore}>
    <WorkspaceSettingsContext.Provider value={[{ editorBinding: EditorBinding.NONE }, jest.fn()]}>
      <MemoryRouter>
        <Dropdown />
      </MemoryRouter>
    </WorkspaceSettingsContext.Provider>
  </Provider>
);

test('Dropdown does not mount Profile, DropdownCourses and DropdownCreateCourses components when a user is not logged in', () => {
  const mockStore = getMockedStore({
    courses: [],
    courseId: 1
  });
  const app = getElement(mockStore);
  const tree = renderTreeJson(app);
  expect(tree).toMatchSnapshot();

  // Expect the Profile component to NOT be mounted
  const dropdown = renderTree(app);
  expect(dropdown.root.findAllByType(Profile).length).toBe(0);
  expect(dropdown.root.findAllByType(DropdownCourses).length).toBe(0);
  expect(dropdown.root.findAllByType(DropdownCreateCourse).length).toBe(0);
});

test('Dropdown correctly mounts Profile, DropdownCourses, and DropdownCreateCourses components when a user is logged in', () => {
  const mockStore = getMockedStore({
    name: 'Some user',
    courses: [],
    courseId: 1
  });
  const app = getElement(mockStore);
  const tree = renderTreeJson(app);
  expect(tree).toMatchSnapshot();

  // Expect the Profile component to be mounted
  const dropdown = renderTree(app);
  expect(dropdown.root.findByType(Profile)).toBeTruthy();
  expect(dropdown.root.findByType(DropdownCourses)).toBeTruthy();
  expect(dropdown.root.findByType(DropdownCreateCourse)).toBeTruthy();
});
