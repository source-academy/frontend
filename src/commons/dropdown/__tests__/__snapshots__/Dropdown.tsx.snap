// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown correctly mounts Profile, DropdownCourses, and DropdownCreateCourses components when a user is logged in 1`] = `
<span
  aria-haspopup="true"
  className="bp4-popover2-target"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <button
    className="bp4-button bp4-minimal"
    disabled={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      aria-hidden={true}
      className="bp4-icon bp4-icon-caret-down"
      icon="caret-down"
    >
      <svg
        data-icon="caret-down"
        height={16}
        role="img"
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>
</span>
`;

exports[`Dropdown does not mount Profile, DropdownCourses and DropdownCreateCourses components when a user is not logged in 1`] = `
<span
  aria-haspopup="true"
  className="bp4-popover2-target"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <button
    className="bp4-button bp4-minimal"
    disabled={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      aria-hidden={true}
      className="bp4-icon bp4-icon-caret-down"
      icon="caret-down"
    >
      <svg
        data-icon="caret-down"
        height={16}
        role="img"
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>
</span>
`;
